presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcAggregatedPageSourceFactory.java

```diff
@@ -0,0 +1,209 @@
+    {
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.functionResolution = requireNonNull(functionResolution, "functionResolution is null");
+        this.useOrcColumnNames = useOrcColumnNames;
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.stats = requireNonNull(stats, "stats is null");
+        this.orcFileTailSource = requireNonNull(orcFileTailSource, "orcFileTailCache is null");
+        this.stripeMetadataSourceFactory = requireNonNull(stripeMetadataSourceFactory, "stripeMetadataSourceFactory is null");
+    }
+
+    @Override
+    public Optional<? extends ConnectorPageSource> createPageSource(
+            Configuration configuration,
+            ConnectorSession session,
+            HiveFileSplit fileSplit,
+            Storage storage,
+            List<HiveColumnHandle> columns,
+            HiveFileContext hiveFileContext,
+            Optional<EncryptionInformation> encryptionInformation,
+            boolean appendRowNumberEnabled)
+    {
+        if (!OrcSerde.class.getName().equals(storage.getStorageFormat().getSerDe())) {
+            return Optional.empty();
+        }
+
+        // per HIVE-13040 and ORC-162, empty files are allowed
+        if (fileSplit.getFileSize() == 0) {
+            return Optional.of(new FixedPageSource(ImmutableList.of()));
+        }
+
+        return Optional.of(createOrcPageSource(
+                session,
+                ORC,
+                hdfsEnvironment,
+                configuration,
+                fileSplit,
+                columns,
+                useOrcColumnNames,
+                typeManager,
+                functionResolution,
+                stats,
+                orcFileTailSource,
+                stripeMetadataSourceFactory,
+                hiveFileContext,
+                encryptionInformation,
+                NO_ENCRYPTION,
+                appendRowNumberEnabled));
+    }
+
+    public static ConnectorPageSource createOrcPageSource(
+            ConnectorSession session,
+            OrcEncoding orcEncoding,
+            HdfsEnvironment hdfsEnvironment,
+            Configuration configuration,
+            HiveFileSplit fileSplit,
+            List<HiveColumnHandle> columns,
+            boolean useOrcColumnNames,
+            TypeManager typeManager,
+            StandardFunctionResolution functionResolution,
+            FileFormatDataSourceStats stats,
+            OrcFileTailSource orcFileTailSource,
+            StripeMetadataSourceFactory stripeMetadataSourceFactory,
+            HiveFileContext hiveFileContext,
+            Optional<EncryptionInformation> encryptionInformation,
+            DwrfEncryptionProvider dwrfEncryptionProvider,
+            boolean appendRowNumberEnabled)
+    {
+        OrcDataSource orcDataSource = getOrcDataSource(session, fileSplit, hdfsEnvironment, configuration, hiveFileContext, stats);
```

Comment:
@rschlussel this is resource leak because we don't close the orcDataSource in a happy case

Reply:
oh good catch. Thank you!